src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -746,18 +753,18 @@ public void close(String reason, boolean flushEngine) throws IOException {
                 }
                 changeState(IndexShardState.CLOSED, reason);
             } finally {
-                final Engine engine = this.currentEngineReference.getAndSet(null);
-                try {
-                    if (flushEngine && this.flushOnClose) {
-                        engine.flushAndClose();
-                    }
-                } finally { // playing safe here and close the engine even if the above succeeds - close can be called multiple times
-                    IOUtils.close(engine);
-                }
+                // we can set the flushOnClose to this value because the only other thing that would update it is the settings refresh
+                // and we removed the listener for setting changes above
+                this.flushOnClose = (flushEngine && this.flushOnClose);
+                decRef();
             }
         }
     }
 
+    void waitForCloseCompletion() throws InterruptedException {
```

Comment:
this should be public to be useful no? I also think we need javadocs what this does and maybe a timeout?